docs/modules/ROOT/pages/adr/0030-function-revisions.adoc

```diff
@@ -0,0 +1,176 @@
+= ADR 0030 - Function Revisions
+:adr_author:    Simon Beck
+:adr_owner:     Schedar
+:adr_reviewers:
+:adr_date:      2025-01-29
+:adr_upd_date:  2025-01-29
+:adr_status:    draft
+:adr_tags:      comp-functions,crossplane
+
+include::partial$adr-meta.adoc[]
+
+[NOTE]
+.Summary
+====
+In order to make rollouts easier, we want to make sure any changes will only be applied to existing instances during their respective maintenance window.
+====
+
+PoC URL: https://github.com/vshn/component-appcat/pull/612
+
+== Context
+Currently, if we have changes that trigger restarts, then we have either to:
+
+* time them during the cluster maintenance
+* disable ArgoCD and manually apply Maintenance hooks to re-enable it again
+
+Both of these solutions are manual and prone to error. Additionally, the first approach breaks our Renovate process that deploys a change on test clusters for a week before its rolled out everywhere.
+
+Crossplane has a concept called https://docs.crossplane.io/latest/concepts/composition-revisions/#using-composition-revisions[Composition Revisions]. They are automatically created each time a change in a composition is applied. By default, any claim or composite deployed will automatically point to the most recent revision as soon as they are available.
+
+This behaviour can be switched to `Manual`, which means that a claim or composite won't automatically update to the latest revision. They are effectively pinned to a version. If a claim or composite gets created with the `Manual` policy, it will point to the latest available revision at the time of creation.
+
+Unfortunately these revisions don't work natively with https://docs.crossplane.io/latest/guides/write-a-composition-function-in-go/[Compostion Functions]. Even though there are `functionRevisions` similar to the `compositionRevisions` they can't be used for anything at the moment.
```

Comment:
Is there a GitHub issue or a design document which discusses functionRevisions? So that we can track what's going on in the Crossplane project regarding that topic. If there is nothing, I suggest opening a GitHub issue once we merged this ADR and point from the GH issue to this ADR, describing how we're currently solving it.

Reply:
I actually opened one, but apart from a few other people being interested in that, nothing much came from it yet: https://github.com/crossplane/crossplane/issues/6139

I'll add the link as well.